classes/webservice/WebserviceOutputBuilder.php

```diff
@@ -659,79 +611,111 @@ protected function renderAssociations($object, $depth, $associations, $ws_params
-                } else {
-                    $output .= $this->setIndent($depth) . $this->objectRender->renderAssociationHeader($object, $ws_params, $assoc_name, true);
-                }
+
+                $this->injectAssociation($associationsWrapperNode, $object, $assoc_name, $association, $ws_params, $objects_assoc, $fields_assoc);
             }
         }
-        $output .= $this->objectRender->renderAssociationWrapperFooter();
+    }
 
-        return $output;
+    /**
+     * Inject one association section into parent node
+     *
+     * @param ApiNode $parentNode
+     * @param ObjectModel $object
+     * @param string $assoc_name
+     * @param array $association
+     * @param array $params
+     * @param array $objects_assoc
+     * @param array|null $fields_assoc
+     *
+     * @return void
+     */
+    private function injectAssociation(ApiNode &$parentNode, ObjectModel $object, string $assoc_name, array $association, array $params, array $objects_assoc, ?array $fields_assoc = null): void
+    {
+        //1) add assocatiation node with attributes
+        $associationNode = $parentNode->addListNode($assoc_name);
+
+        if ($this->schemaToDisplay == 'blank') {//if this is blank schema, adding parent node is sufficient
+            return;
+        }
+
+        $associationNode->addAttribute('nodeType', $params['associations'][$assoc_name]['resource']);
+
+        if (array_key_exists('setter', $params['associations'][$assoc_name]) && !$params['associations'][$assoc_name]['setter']) {
+            $associationNode->addAttribute('readOnly', 'true');
+        }
+
+        if (isset($params['associations'][$assoc_name]['virtual_entity']) && $params['associations'][$assoc_name]['virtual_entity']) {
+            $associationNode->addAttribute('virtualEntity', 'true');
+        } else {
+            if (isset($params['associations'][$assoc_name]['api'])) {
+                $associationNode->addAttribute('api', $params['associations'][$assoc_name]['api']);
+            } else {
+                $associationNode->addAttribute('api', $assoc_name);
+            }
+        }
+
+        //2) Add details underneath association node
+
+        $parent_details = [
+            'object_id' => $object->id,
+            'entity_name' => $params['objectNodeName'],
+            'entities_name' => $params['objectsNodeName'],
+        ];
+
+        foreach ($objects_assoc as $object_assoc) {
+            $value = $object_assoc ?? null;
+            $fields_assoc = empty($fields_assoc) ? [['id' => $value['id']]] : $fields_assoc;
+            $this->injectFlatAssociation($associationNode, $object, $assoc_name, $association['resource'], $fields_assoc, $value, $parent_details);
+        }
     }
 
-    protected function renderFlatAssociation($object, $depth, $assoc_name, $resource_name, $fields_assoc, $object_assoc, $parent_details)
```

Comment:
Could you document the BC Break ? (Removal of this method)

Reply:
If its not being called anywhere anymore, its not a BC break, no?